filebeat/input/journald/pkg/journalctl/journalctl.go

```diff
@@ -0,0 +1,139 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package journalctl
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+
+	input "github.com/elastic/beats/v7/filebeat/input/v2"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type journalctl struct {
+	cmd      *exec.Cmd
+	dataChan chan []byte
+	stdout   io.ReadCloser
+	stderr   io.ReadCloser
+
+	logger   *logp.Logger
+	canceler input.Canceler
+}
+
+// Factory returns an instance of journalctl ready to use.
+//
+// The returned type is an interface to allow mocking for testing
+func Factory(canceller input.Canceler, logger *logp.Logger, binary string, args ...string) (Jctl, error) {
+	cmd := exec.Command(binary, args...)
+
+	jctl := journalctl{
+		canceler: canceller,
+		cmd:      cmd,
+		dataChan: make(chan []byte),
+		logger:   logger,
+	}
+
+	var err error
+	jctl.stdout, err = cmd.StdoutPipe()
+	if err != nil {
+		return &journalctl{}, fmt.Errorf("cannot get stdout pipe: %w", err)
+	}
+	jctl.stderr, err = cmd.StderrPipe()
+	if err != nil {
+		return &journalctl{}, fmt.Errorf("cannot get stderr pipe: %w", err)
+	}
+
+	go func() {
+		defer jctl.logger.Debug("stderr reader goroutine done")
+		reader := bufio.NewReader(jctl.stderr)
+		for {
+			line, err := reader.ReadString('\n')
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					return
+				}
+				logger.Errorf("cannot read from journalctl stderr: %s", err)
+				return
+			}
+
+			logger.Errorf("Journalctl wrote to stderr: %s", line)
+		}
+	}()
+
+	// Goroutine to read events from stdout
+	go func() {
+		defer jctl.logger.Debug("stdout reader goroutine done")
+		defer close(jctl.dataChan)
+		reader := bufio.NewReader(jctl.stdout)
+		for {
+			data, err := reader.ReadBytes('\n')
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					return
+				}
+				logger.Errorf("cannot read from journalctl stdout: %s", err)
+				return
+			}
+
+			select {
+			case <-jctl.canceler.Done():
+				return
+			case jctl.dataChan <- data:
+			}
+		}
+	}()
+
+	logger.Infof("Journalctl command: journalctl %s", strings.Join(args, " "))
+
+	if err := cmd.Start(); err != nil {
+		return &journalctl{}, fmt.Errorf("cannot start journalctl: %w", err)
+	}
+
+	logger.Infof("journalctl started with PID %d", cmd.Process.Pid)
+
+	go func() {
+		if err := cmd.Wait(); err != nil {
+			jctl.logger.Errorf("journalctl exited with an error, exit code %d ", cmd.ProcessState.ExitCode())
+		}
+	}()
+
+	return &jctl, nil
+}
+
+// Kill Terminates the journalctl process using a SIGKILL.
+func (j *journalctl) Kill() error {
+	j.logger.Debug("sending SIGKILL to journalctl")
+	err := j.cmd.Process.Kill()
+	return err
+}
+
+func (j *journalctl) Next(cancel input.Canceler) ([]byte, error) {
+	select {
+	case <-cancel.Done():
```

Comment:
if cancel.Done(), do we have to kill the process?

Reply:
This `input.Canceler` is in the context of the `Next` call, so I don't think we should kill the journalctl process. If the journald input is stopped, it will call Close on the Reader:

https://github.com/belimawr/beats/blob/e7d226030d8a90477949a4cd3a013e09c28db519/filebeat/input/journald/input.go#L236-L238

and the reader will kill the journalctl process

https://github.com/belimawr/beats/blob/e7d226030d8a90477949a4cd3a013e09c28db519/filebeat/input/journald/pkg/journalctl/reader.go#L192-L200.

I added some comments explaining kill needs to be called.